File: backend/app/schemas/scoreboard.py
"""Response models for the scoreboard.

Entries stay loosely typed (`dict[str, Any]`) because both boards are computed
once, cached whole, and served from that projection — re-validating every row
through a model on the way out would buy nothing and would mean two places to
change when a column is added.

What *is* enforced is the exclusion: `scoreboard_cache.public_view` strips
`score` before either board is served, and a test asserts no public response
carries it (spec 059 §2).
"""

from typing import Any

from pydantic import BaseModel


class PlayerBoardResponse(BaseModel):
    total: int
    generated_at: str
    entries: list[dict[str, Any]]


class TeamBoardResponse(BaseModel):
    total: int
    generated_at: str
    entries: list[dict[str, Any]]


class MyStandingResponse(BaseModel):
    """A compact header: where you stand, and where your party stands.

    No XP, for §2's reason — this is a board surface, and a party's total sitting
    on an endpoint nobody currently renders is exactly the field that gets
    rendered by accident later. Level is the public shape of the same thing.
    """

    rank: int | None
    level: int
    player_count: int
    team_rank: int | None
    team_level: int | None
    team_count: int


class BossStarResponse(BaseModel):
    """One boss kill, identified by the challenge's slug (spec 059 §3)."""

    slug: str
    tier: str
    #: 1 (Neighborhood) to 6 (Floor), so a client orders without the names.
    level: int
    #: For the hover. Colour carries the tier; this carries which boss.
    title: str


class PartyMemberResponse(BaseModel):
    user_id: str
    display_name: str
    has_avatar: bool
    level: int
    class_name: str | None
    class_rarity: str | None


class PartyPanelResponse(BaseModel):
    """Who a party is (spec 059 §5). Deliberately no XP, member XP included."""

    team_id: str
    name: str
    rank: int
    level: int
    member_count: int
    #: Distinct challenges solved by any current member.
    solve_count: int
    #: Distinct achievements held by current members.
    achievement_count: int
    stars: list[BossStarResponse]
    founded_at: str
    members: list[PartyMemberResponse]


class ActivityItemResponse(BaseModel):
    """One line of the ticker (spec 069).

    `challenge_title` and `tier` are null unless `kind` is "boss" — §3's decision
    enforced on the server, because a title the client is asked to hide is a
    title in the payload.
    """

    kind: str
    display_name: str
    zone_name: str
    challenge_title: str | None
    tier: str | None
    tier_level: int | None
    at: str


class ActivityResponse(BaseModel):
    items: list[ActivityItemResponse]

File: frontend/src/api/scoreboard.ts
import { api } from "./client";

/**
 * The boards (spec 005, rebuilt by 059).
 *
 * **Neither entry has a `score`.** XP still orders both boards server-side; it
 * is not sent, and adding it back here is how it ends up rendered. The one place
 * XP belongs is the player's own character sheet.
 */

export type BossTier =
  | "neighborhood"
  | "borough"
  | "city"
  | "province"
  | "country"
  | "floor";

/** One boss kill, identified by the challenge's slug (spec 059 §3). */
export interface BossStar {
  slug: string;
  tier: BossTier;
  /** 1 (Neighborhood) to 6 (Floor), so a run orders without the names. */
  level: number;
  /** For the hover. Colour carries the tier; this carries which boss. */
  title: string;
}

export interface PlayerEntry {
  /** Shared on a tie, so a board can show two third places. */
  rank: number;
  user_id: string;
  display_name: string;
  has_avatar: boolean;
  team_id: string | null;
  team_name: string | null;
  level: number;
  solve_count: number;
  last_gain_at: string | null;
  /** This player's own kills — never their party's. */
  stars: BossStar[];
  /** The worn loot title (038) and the class (016/024). Both cosmetic. */
  title: string | null;
  class_name: string | null;
  class_rarity: string | null;
}

export interface TeamEntry {
  rank: number;
  team_id: string;
  name: string;
  member_count: number;
  level: number;
  /** Distinct challenges solved by any current member. */
  solve_count: number;
  last_gain_at: string | null;
  /** The union across current members, deduplicated by slug. */
  stars: BossStar[];
}

export interface Boards {
  generated_at: string;
  players: PlayerEntry[];
  teams: TeamEntry[];
}

export interface MyStanding {
  rank: number | null;
  level: number;
  player_count: number;
  team_rank: number | null;
  team_level: number | null;
  team_count: number;
}

export interface PartyMember {
  user_id: string;
  display_name: string;
  has_avatar: boolean;
  level: number;
  class_name: string | null;
  class_rarity: string | null;
}

/** Who a party is (spec 059 §5). No XP, member XP included. */
export interface PartyPanel {
  team_id: string;
  name: string;
  rank: number;
  level: number;
  member_count: number;
  solve_count: number;
  achievement_count: number;
  stars: BossStar[];
  founded_at: string;
  members: PartyMember[];
}

/**
 * All the rows in one request.
 *
 * The top-ten framing, the search and "show all" are client-side on a payload
 * the server already computes whole, so paging it would mean asking for
 * something it has already sent (spec 059 §4). 500 is the endpoint's ceiling.
 */
const WHOLE_BOARD = "?limit=500";

export const getPlayerBoard = () =>
  api.get<{ total: number; generated_at: string; entries: PlayerEntry[] }>(
    `/scoreboard/players${WHOLE_BOARD}`,
  );

export const getTeamBoard = () =>
  api.get<{ total: number; generated_at: string; entries: TeamEntry[] }>(
    `/scoreboard/teams${WHOLE_BOARD}`,
  );

export const getMyStanding = () => api.get<MyStanding>("/scoreboard/me");

export const getPartyPanel = (teamId: string) =>
  api.get<PartyPanel>(`/scoreboard/teams/${teamId}`);

/** The socket rides the same origin as the API, so the session cookie goes with it. */
export function scoreboardSocketUrl(): string {
  const protocol = window.location.protocol === "https:" ? "wss:" : "ws:";
  return `${protocol}//${window.location.host}/api/ws/scoreboard`;
}
